tools/src/wyvern/tools/typedAST/core/declarations/ConstructDeclaration.java

```diff
@@ -0,0 +1,126 @@
+import wyvern.tools.errors.FileLocation;
+import wyvern.tools.typedAST.abs.Declaration;
+import wyvern.tools.typedAST.core.binding.NameBinding;
+import wyvern.tools.typedAST.core.binding.NameBindingImpl;
+import wyvern.tools.typedAST.interfaces.CoreAST;
+import wyvern.tools.typedAST.interfaces.ExpressionAST;
+import wyvern.tools.typedAST.interfaces.TypedAST;
+import wyvern.tools.types.Type;
+import wyvern.tools.types.UnresolvedType;
+
+public class ConstructDeclaration extends Declaration implements CoreAST {
+    private ExpressionAST definition;
+    private NameBinding binding;
+    private List<NameBinding> argNames;
+
+    public ConstructDeclaration(String name, List<NameBinding> argNames, TypedAST definition, FileLocation location) {
+        this.definition = (ExpressionAST) definition;
+        binding = new NameBindingImpl(name, null);
+        this.argNames = argNames;
+        this.location = location;
+    }
+
+    // @Override
+    // public List<NameBinding> getArgBindings() {
+    //     return argNames;
+    // }
+
+    @Override
+    public Type getType() {
+        return binding.getType();
+    }
+
+    @Override
+    public String getName() {
+        return binding.getName();
+    }
+
+    public ExpressionAST getDefinition() {
+        return definition;
+    }
+
+    private FileLocation location = FileLocation.UNKNOWN;
+    public FileLocation getLocation() {
+        return this.location; //TODO
+    }
+
+    @Override
+    public void genTopLevel(TopLevelContext tlc) {
+        ValueType declType = getILValueType(tlc.getContext());
+        tlc.addLet(getName(), getILValueType(tlc.getContext()), definition.generateIL(tlc.getContext(), declType, tlc.getDependencies()), false);
+    }
+
+    @Override
+    public DeclType genILType(GenContext ctx) {
+        ValueType vt = getILValueType(ctx);
+        return new ValDeclType(getName(), vt);
+    }
+
+    private ValueType getILValueType(GenContext ctx) {
+        ValueType vt;
+        final Type type = new UnresolvedType("Unit", FileLocation.UNKNOWN);
+        vt = type.getILType(ctx);
+        return vt;
+    }
+
+    @Override
+    public wyvern.target.corewyvernIL.decl.Declaration generateDecl(GenContext ctx, GenContext thisContext) {
+
+        ValueType expectedType = getILValueType(thisContext);
+        /* uses ctx for generating the definition, as the selfName is not in scope */
+
+        List<FormalArg> args = new LinkedList<FormalArg>();
+        GenContext methodContext = thisContext;
+        for (NameBinding b : argNames) {
+            ValueType argType = b.getType().getILType(thisContext);
+            args.add(new FormalArg(b.getName(), argType));
+            methodContext = methodContext.extend(b.getName(), new Variable(b.getName()), argType);
+            thisContext = thisContext.extend(b.getName(), new Variable(b.getName()), argType);
+        }
+
+        //need to create ConstructDeclaration which accepts args
```

Comment:
As we discussed, this should generate a ValDeclaration if there are no arguments, and a DefDeclaration if there are arguments.  We also need to generate a type declaration for each constructor, but that probably has to go in a separate method, since this method can only return one declaration.

Reply:
Anyway this doesn't have to be fixed right away as we discussed.  Fix the other issues now though, please!